### PR TITLE
#1214 #1176 Build Transaction bottleNeck

### PR DESCRIPTION
**Pull request by kogot**
Used memory cache and SecureString for secretKey with 5 min cache clean up timer.

### CODE REVIEW

File: src/Stratis.Bitcoin/Utilities/Extensions/StringExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Security;
+using System.Text;
+
+namespace Stratis.Bitcoin.Utilities.Extensions
+{
+    public static class StringExtensions
+    {
+        public static SecureString ToSecureString(this string input)
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Utilities/Extensions/StringExtensions.cs]: Should you not create a new method `FromSecureString()`?

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 2]
```diff
@@ -208,9 +213,21 @@ private void AddSecrets(TransactionBuildContext context)
                 return;
 
             Wallet wallet = this.walletManager.GetWalletByName(context.AccountReference.WalletName);
-
+            Key privateKey;
             // get extended private key
-            var privateKey = Key.Parse(wallet.EncryptedSeed, context.WalletPassword, wallet.Network);
+            string cacheKey = $"{wallet.EncryptedSeed}{context.WalletPassword}{wallet.Network}";
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: `context.WalletPassword` should be a secure string on its own.
And not a good idea to use the password as cash key that will make it vulnerable as the password can easily extract the key.

File: src/Stratis.Bitcoin/Utilities/Extensions/StringExtensions.cs (hunk 1 above)
Comment (kogot) [reply src/Stratis.Bitcoin/Utilities/Extensions/StringExtensions.cs]: Good point.

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 2 above)
Comment (kogot) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Well, I took it from your conversation on issue #1214 , If I won't use it as combined key, than anyone can fake cacheKey and get Private Key from cache, or not? @dangershony

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: I think retrieving from cache is not that hard anyway even without the password, the point of secure string is that the content is encrypted in memory (and if cache is persisted to disk) so by keeping the password in clear memory we expose the key (as also the encrypted seed is in clear memory.

We need to keep the password as SecureString as well, the only pint where the password/private seed can be not as SecureString is at the point of deriving the prv key and the point we read it form the controller.

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 3]
```diff
@@ -38,6 +41,8 @@ public class WalletTransactionHandler : IWalletTransactionHandler
 
         private readonly ILogger logger;
 
+        private MemoryCache privateKeyCache = new MemoryCache(new MemoryCacheOptions(){ExpirationScanFrequency = new TimeSpan(0,1,0)});
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Please initialize this in the constructor, it's easier when debugging :-)

Comment (kogot) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Done

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 4]
```diff
@@ -208,9 +215,22 @@ private void AddSecrets(TransactionBuildContext context)
                 return;
 
             Wallet wallet = this.walletManager.GetWalletByName(context.AccountReference.WalletName);
-
+            Key privateKey; 
+            SHA256 sha256 = SHA256.Create();
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: You can use Hashes.Hash256().

File: src/Stratis.Bitcoin/Utilities/Extensions/StringExtensions.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Utilities/Extensions/StringExtensions.cs]: Just one more thing please add comments

**Comment by dangershony** [comment]
Two more comments and good to go.

**Comment by quantumagi** [comment]
I have some extra comments and a FromSecureString method in my file on PR 1419 so feel free to copy that code:

https://github.com/stratisproject/StratisBitcoinFullNode/pull/1419/files#diff-d78f297655de814db43b6100aa5d07a4

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 5]
```diff
@@ -208,9 +215,22 @@ private void AddSecrets(TransactionBuildContext context)
...
             // get extended private key
-            var privateKey = Key.Parse(wallet.EncryptedSeed, context.WalletPassword, wallet.Network);
+            string cacheKey = Encoding.Default.GetString(sha256.ComputeHash(Encoding.UTF8.GetBytes(wallet.EncryptedSeed)));
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Why is the cacheKey created in his way? It's just an index for the seed in the MemoryCache, right? (I spent 5 minutes trying to understand why it's hashed, encoded etc.)
I prefer the simple approach of doing `string cacheKey = wallet.EncryptedSeed;`.
Would that work?

**Comment by kogot** [comment]
@dangershony @bokobza done

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 6]
```diff
@@ -208,9 +216,21 @@ private void AddSecrets(TransactionBuildContext context)
                 return;
 
             Wallet wallet = this.walletManager.GetWalletByName(context.AccountReference.WalletName);
-
+            Key privateKey; 
             // get extended private key
-            var privateKey = Key.Parse(wallet.EncryptedSeed, context.WalletPassword, wallet.Network);
+            string cacheKey = Encoding.Default.GetString(Hashes.Hash256(Encoding.UTF8.GetBytes(wallet.EncryptedSeed)).ToBytes());
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: @kogot I still don't understand why is the cackeKey calculated in this way. 😕

Comment (kogot) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: @bokobza yeah reason was mostly to hide plain value cause we storing it in memory. If you so unhappy with that, I'm easy to use EncryptedSeed as key as is.

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: I think it's ok. The encrypted seed is encrypted so no problem with that IMO.

**Comment by bokobza** [comment]
That looks good to me now. 
@kogot have you tried sending some transactions to test it works?
Are the Integration tests passing ok as well?

**Pull request merged**